File: fixtures/capability/adapters/node/adapter.ts
// Reference capability adapter (Node). Implements the protocol decision rules from the
// published contract alone. Must agree with the Python and Go adapters; any disagreement
// means the contract is not language-neutral and ARC-006 is unmet.
import { readFileSync } from "node:fs";

interface Contract {
  readonly supportedMajors: readonly number[];
  readonly requiredManifestKeys: readonly string[];
  readonly audience: string;
}
interface Case {
  readonly name: string;
  readonly kind: string;
  readonly now: number;
  readonly payload?: Record<string, unknown>;
}
interface Verdict { case: string; accepted: boolean; code: string | null }

const contractPath = process.argv[2] as string;
const corpusPath = process.argv[3] as string;
const contract = JSON.parse(readFileSync(contractPath, "utf8")) as Contract;
const corpus = (JSON.parse(readFileSync(corpusPath, "utf8")) as { cases: Case[] }).cases;

function decide(item: Case): string | null {
  const payload = item.payload ?? {};
  const version = payload.contractVersion as { major?: unknown } | undefined;
  if (!version || typeof version.major !== "number") return "MALFORMED_MANIFEST";
  if (!contract.supportedMajors.includes(version.major)) return "UNSUPPORTED_CONTRACT_VERSION";
  if (item.kind === "manifest") {
    for (const key of contract.requiredManifestKeys) {
      if (!Object.prototype.hasOwnProperty.call(payload, key)) return "MALFORMED_MANIFEST";
    }
    const signature = payload.signature;
    if (typeof signature !== "string" || signature.length === 0) return "UNSIGNED_MANIFEST";
  }
  if (item.kind === "principal") {
    if (payload.aud !== contract.audience) return "UNAUTHORIZED_CAPABILITY";
    const exp = payload.exp;
    if (typeof exp !== "number" || exp <= item.now) return "UNAUTHORIZED_CAPABILITY";
  }
  return null;
}

const out: Verdict[] = corpus.map((item) => {
  const code = decide(item);
  return { case: item.name, accepted: code === null, code };
});
process.stdout.write(JSON.stringify(out));

File: fixtures/capability/adapters/python/adapter.py
"""Reference capability adapter (Python). Mirrors the Node and Go adapters."""
import json
import sys

contract_path, corpus_path = sys.argv[1], sys.argv[2]
with open(contract_path, encoding="utf-8") as handle:
    contract = json.load(handle)
with open(corpus_path, encoding="utf-8") as handle:
    corpus = json.load(handle)["cases"]


def decide(item):
    payload = item.get("payload") or {}
    version = payload.get("contractVersion")
    if not isinstance(version, dict) or not isinstance(version.get("major"), int) or isinstance(version.get("major"), bool):
        return "MALFORMED_MANIFEST"
    if version["major"] not in contract["supportedMajors"]:
        return "UNSUPPORTED_CONTRACT_VERSION"
    if item.get("kind") == "manifest":
        for key in contract["requiredManifestKeys"]:
            if key not in payload:
                return "MALFORMED_MANIFEST"
        signature = payload.get("signature")
        if not isinstance(signature, str) or signature == "":
            return "UNSIGNED_MANIFEST"
    if item.get("kind") == "principal":
        if payload.get("aud") != contract["audience"]:
            return "UNAUTHORIZED_CAPABILITY"
        exp = payload.get("exp")
        if not isinstance(exp, int) or isinstance(exp, bool) or exp <= item["now"]:
            return "UNAUTHORIZED_CAPABILITY"
    return None


out = []
for entry in corpus:
    code = decide(entry)
    out.append({"case": entry["name"], "accepted": code is None, "code": code})
sys.stdout.write(json.dumps(out, separators=(",", ":")))
